daemon: add access checkers needed for /v2/system-volumes endpoint

ZeyadYasser opened the pull request:
This PR adds the following:
- `byActionAccess` access checker decides which access checker to use based on the JSON `action` field `{"action": "<action>"}` to allow using different access checkers based on the target action.
  -  This is intended to be used by the `/v2/system-volumes` endpoint but can also be used for similar action-based endpoints if needed.
- `interfaceRootAccess` access checker to allow root requests over `snapd-snap.socket` for snaps that have a connection of specific interface and are present on the plug side of that connection.
- add optional polkit support to root* access checkers

Example of how those new access checkers would be used:
```go
var systemVolumesCmd = &Command{
	Path: "/v2/system-volumes",
	GET:  getSystemVolumes,
	POST: postSystemVolumesAction,
	ReadAccess: interfaceOpenAccess{
		Interfaces: []string{"snap-fde"},
	},
	WriteAccess: byActionAccess{
		ByAction: map[string]accessChecker{
			"generate-recovery-key": interfaceRootAccess{
				Polkit:     "io.snapcraft.snapd.manage-fde",
				Interfaces: []string{"snap-fde"},
			},
			"replace-recovery-key": interfaceRootAccess{
				Polkit:     "io.snapcraft.snapd.manage-fde",
				Interfaces: []string{"snap-fde"},
			},
			"check-passphrase": interfaceOpenAccess{
				Interfaces: []string{"snap-fde"},
			},
		},
		Default: rootAccess{},
	},
}
```

pedronis commented on daemon/access.go:
```diff
@@ -143,6 +148,14 @@ func (ac rootAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, us
+	// We check polkit last because it may result in the user
+	// being prompted for authorisation. This should be avoided if
+	// access is otherwise granted.
+	if ac.Polkit != "" {
```
that's how the other Polkit field work though, I agreed we need to change the doc comment for the accessor

pedronis commented on daemon/access.go:
```diff
@@ -314,9 +328,95 @@ func (ac interfaceProviderRootAccess) CheckAccess(d *Daemon, r *http.Request, uc
+	Polkit     string
+}
+
+func (ac interfaceRootAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, user *auth.UserState) *apiError {
```
I think we want as few knobs as possible per accessor, basically I think having a single accessor with a lot of config is an anti-goal, we could share implementation via a helper though possibly

olivercalder replied:
I agree, now that `rootAccess` has a polkit authorization as well, it's even harder to keep track of which checker allows access in which ways. I think some sort of matrix, either as documentation or via a helper method which other checkers call, would really help clarify this.

I think it should be easily discoverable if a checker allows access to snapd-snap.socket, requires root ucred UID, authenticated user, polkit, particular interfaces, etc. Ideally one or a small number (if there are several distinct "categories") of helpers which are called by the real checkers could help clarify this. But perhaps that would be too complicated if different checkers have conflicting order of checks or short-circuit in different ways.

Regardless, I think it would be nice if the comments for each type of checker which allows polkit authentication mentioned polkit and the order of checks/short circuits. For example, polkit authentication overrides root requirement but doesn't override snapd.socket requirement for `rootAccess`, but for `interfaceRootAccess` it overrides the root requirement but does not override the interface requirement.

ZeyadYasser replied:
Thanks for the suggestions, I added a common helper and indeed the checkers logic is less confusing now. also much less error-prone repetition.

pedronis reviewed (commented):
couple of initial comments on top of Maciej's input

olivercalder commented on daemon/access_test.go:
```diff
@@ -214,6 +215,40 @@ func (s *accessSuite) TestRootAccess(c *C) {
+	user := &auth.UserState{}
+	ucred := &daemon.Ucrednet{Uid: 0, Pid: 100, Socket: dirs.SnapdSocket}
+
+	// polkit is not checked if any of:
+	//   * ucred is missing
+	//   * user is root
```
This kind of list would be nice to have in the `rootAccess` doc comment as well, something like:

```go
// Access is allowed if:
// - socket is snapd.socket AND
//   - user is root OR
//   - polkit authentication is used
```

ZeyadYasser replied:
I update the doc comments to be consistent with the existing ones, let me know if you think it needs to be changed

olivercalder commented on daemon/access_test.go:
```diff
@@ -601,6 +636,309 @@ plugs:
+	req := httptest.NewRequest("GET", "/", nil)
+	user := &auth.UserState{}
+
+	// polkit is not checked if any of:
+	//   * ucred is missing
+	//   * user is root
+	//   * snap request (as root) with relevant connected slot
+	//   * snap request without relevant connected slot
```
Similar here, would be useful in the doc comment

olivercalder reviewed (commented):
Changes look good, but I agree with Maciej and Samuele comments around being careful with reading from the incoming request.

More generally though, I think the relationships between the different ways to get access for any given checker are increasingly complex. Ideally there would be some way to easily tell how each checker differs and when to use what. This could be a comment with matrix of access and sockets etc., or even a helper function if possible. That may be better suited for a follow-up though, since it's not a new problem, just slightly exacerbated by the additional polkit access for e.g. `rootAccess`.

ZeyadYasser commented on daemon/access.go:
```diff
@@ -314,9 +328,95 @@ func (ac interfaceProviderRootAccess) CheckAccess(d *Daemon, r *http.Request, uc
+	bodyBuf := new(bytes.Buffer)
+	checkerBuf := new(bytes.Buffer)
+	w := io.MultiWriter(bodyBuf, checkerBuf)
+	if _, err := io.Copy(w, r.Body); err != nil {
```
thanks, added

ZeyadYasser commented on daemon/access.go:
```diff
@@ -251,70 +309,150 @@ type interfaceOpenAccess struct {
+	//
+	// Note: The specified polkit action must require auth_admin
+	// to avoid compromising security.
+	Polkit string
```
just for completeness and consistency, same for `rootAccess`

pedronis replied:
I would leave it out for now

ZeyadYasser replied:
Done, removed `Polkit` from `rootAccess` and `interfaceProviderRootAccess` for now as they are not needed. Also, updated doc comment for `interfaceRootAccess` to mention that it supports polkit unlike `rootAccess`.

pedronis reviewed (commented):
gave direction on points I opened in the last review

bboozzoo commented on daemon/access.go:
```diff
@@ -251,47 +299,39 @@ type interfaceOpenAccess struct {
-		Interfaces: ac.Interfaces,
-		Plug:       true,
-	})
+	opts := accessOptions{
+		accessLevel: accessLevelOpen,
+		interfaceAccess: &interfaceAccessReqs{
+			Interfaces: ac.Interfaces,
+			Plug:       true,
+		},
+	}
```
shouldn't this require snap socket?

bboozzoo commented on daemon/access.go:
```diff
@@ -77,25 +81,85 @@ type accessChecker interface {
+		}
+	}
+
+	if opts.accessLevel == accessLevelOpen {
```
we should likely error out on `accessLevel == ""`

jslarraz commented on daemon/access.go:
```diff
@@ -77,25 +81,85 @@ type accessChecker interface {
+	}
+
+	if opts.interfaceAccess != nil {
+		rspe := requireInterfaceApiAccess(d, r, ucred, *opts.interfaceAccess)
```
This check only applies if the request was received via snapd-snap.socket. It might be worth adding a commentary to make it evident or expand the condition `if opts.interfaceAccess != nil && ucred.Socket != dirs.SnapSocket`.

jslarraz commented on daemon/access.go:
```diff
@@ -251,47 +299,39 @@ type interfaceOpenAccess struct {
+	return checkAccess(d, r, ucred, user, opts)
 }
 
 // interfaceAuthenticatedAccess behaves like authenticatedAccess, but
```
```suggestion
// interfaceAuthenticatedAccess behaves like authenticatedAccess, but also
```

jslarraz commented on daemon/access.go:
```diff
@@ -302,19 +342,106 @@ type interfaceProviderRootAccess struct {
+type byActionAccess struct {
+	// ByAction maps from detected request action to access checker.
+	ByAction map[string]accessChecker
+	// Default is the fallback access checker if no action was matched
```
Is there any known case where a valid requests content may not be `json` or include an unknown action (i.e. it was not added to the ByAction field before hand)? Otherwise, I think we should consider returning BadRequest on these situations.

olivercalder commented on daemon/access.go:
```diff
@@ -279,32 +314,15 @@ type interfaceAuthenticatedAccess struct {
-	}
-
-	return Unauthorized("access denied")
+	opts := accessOptions{
+		accessLevel: accessLevelAuthenticated,
+		interfaceAccess: &interfaceAccessReqs{
+			Interfaces: ac.Interfaces,
+			Plug:       true,
+		},
+		polkitAction: ac.Polkit,
+	}
```
Shouldn't this also require `dirs.SnapSocket`?

ZeyadYasser replied:
not really, it meant to extend access to access not make it exclusive unlike `snapAccess` which is exclusive to snaps and is only used for the snapctl api.

```
./api_prompting.go:59:          WriteAccess: interfaceAuthenticatedAccess{Interfaces: []string{"snap-interfaces-requests-control"}, Polkit: polkitActionManage},
./api_prompting.go:67:          WriteAccess: interfaceAuthenticatedAccess{Interfaces: []string{"snap-interfaces-requests-control"}, Polkit: polkitActionManage},
./api_apps.go:49:               WriteAccess: interfaceAuthenticatedAccess{Interfaces: []string{"ros-snapd-support"}, Polkit: polkitActionManage},
./api_themes.go:49:             WriteAccess: interfaceAuthenticatedAccess{Interfaces: []string{"snap-themes-control"}, Polkit: polkitActionManage},
```

olivercalder commented on daemon/access.go:
```diff
@@ -316,26 +334,15 @@ type interfaceProviderRootAccess struct {
+			Slot:       true,
+		},
+		polkitAction: ac.Polkit,
+	}
```
Is it worth specifying both sockets as explicitly allowed here?

olivercalder commented on daemon/access.go:
```diff
@@ -347,26 +354,15 @@ type interfaceRootAccess struct {
-
-	return Unauthorized("access denied")
+	opts := accessOptions{
+		accessLevel: accessLevelRoot,
+		interfaceAccess: &interfaceAccessReqs{
+			Interfaces: ac.Interfaces,
+			Plug:       true,
+		},
+		polkitAction: ac.Polkit,
```
Same, explicitly list both sockets?

jslarraz commented on daemon/access.go:
```diff
@@ -302,19 +369,108 @@ type interfaceProviderRootAccess struct {
+
+	checker := ac.ByAction[req.Action]
+	if checker == nil {
+		return ac.Default.CheckAccess(d, r, ucred, user)
```
What about unknown actions? Is there any supported case where actions are not know before hand so that they can be added to  ByAction map? If not, I think we should consider also returning BadRequest in this case rather than using a default checker

pedronis reviewed (commented):
unsure about forcing to enumerate actions

ZeyadYasser commented:
@pedronis @jslarraz I reverted https://github.com/canonical/snapd/pull/15638/commits/df111be9df445fac0ba918400a32f4d2159b9c54 and added checks that the default access checker must be root-level.

jslarraz reviewed (approved):
LGTM
